changefeedccl: add on_error option to pause changefeeds on failure

spiffyy99 opened the pull request:
Previously, changefeeds always failed when encountering a non-
retryable error. This option allows the user to pause on failure
and resume later, while still failing as default behavior.

Release note (enterprise change): new 'on_error' option to pause
on non-retryable errors instead of failing.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68176)
<!-- Reviewable:end -->

spiffyyeng commented on pkg/ccl/changefeedccl/changefeed_stmt.go:
```diff
@@ -665,8 +709,6 @@ func (b *changefeedResumer) Resume(ctx context.Context, execCtx interface{}) err
 		}
 	}
-	// We only hit this if `r.Next()` returns false, which right now only happens
-	// on context cancellation.
```
outdated comment: we can hit this if the retry count is exhausted as well

miretskiy reviewed (changes requested):
Very nice!
Just few minor nits/questions.

> Reviewed 6 of 6 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-:-MfiZzrR9EYTkM4c47pr:b15sb22)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 555 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiUv1l7mlo8YvuZ8li:-MfiUv1l7mlo8YvuZ8lj:b-np852r) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L555)):*
> ```Go
> 		default:
> 			return jobspb.ChangefeedDetails{}, errors.Errorf(
> 				`unknown %s: %s`, opt, v)
> ```

I would add maybe a a hint (errors.WithHint), or just expand the error message with "valid values are....."
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiW3Ng7o2axOiaOVho:-MfiW3Ng7o2axOiaOVhp:b-cgauf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L629)):*
> ```Go
> 				return pauseErr
> 			}
> 			// TODO (ryan min): Populate pause reason with error once column is added (#67928)
> ```

We already have that.  See running status (see changefeed_processors).
In order to update run status and indicate that we're pausing because of this policy, as opposed to user initiated action, we will probably need to export `pauseRequested` method  (as well as `onPauseRequestFunc`) so that we can directly call b.job.PauseRequested, and update run status in the passed in function.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 633 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiVcVaBOVWaSiTse4G:-MfiVcVaBOVWaSiTse4H:baogisp) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L633)):*
<details><summary><i alt="summary">Quoted 18 lines of code…</i></summary>

> ```
> 		switch onError := changefeedbase.OnErrorType(details.Opts[changefeedbase.OptOnError]); onError {
> 		// default behavior
> 		case changefeedbase.OptOnErrorFail:
> 			return err
> 		// pause instead of failing
> 		case changefeedbase.OptOnErrorPause:
> 			// note: we only want the job to pause here if a failure happens, not a
> 			// user-initiated cancellation. if the job has been canceled, the ctx
> 			// will handle it and the pause will return an error.
> 			pauseErr := execCfg.JobRegistry.PauseRequested(ctx, jobExec.ExtendedEvalContext().Txn, jobID)
> 			if pauseErr != nil {
> 				return pauseErr
> 			}
> 			// TODO (ryan min): Populate pause reason with error once column is added (#67928)
> 		default:
> 			return errors.Errorf("unrecognized option value: %s=%s",
> 				changefeedbase.OptOnError, details.Opts[changefeedbase.OptOnError])
> 		}
> ```
</details>


I would move this into a helper `handleChangefeedError()` which returns an error (or nil)..
So, this code becomes:
```
if err != nil {
   err = handleChangefeedError(....)
}
return err
```
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiYjiE5qX2Ci44amLY:-MfiYjiE5qX2Ci44amLZ:b-f38s62) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_test.go#L4265)):*
> ```Go
> 
> 			// check for paused status on failure
> 			err := feedJob.WaitForStatus(func(s jobs.Status) bool { return s == jobs.StatusPaused })
> ```

require.NoError(t, feedJob.WaitForStatus...)?
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4345 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiYpsw6ZM-3VQkShCL:-MfiYpsw6ZM-3VQkShCM:b-j6rd9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_test.go#L4345)):*
> ```Go
> 	t.Run(`cloudstorage`, cloudStorageTest(testFn))
> 	t.Run(`kafka`, kafkaTest(testFn))
> 	t.Run(`webhook`, webhookTest(testFn))
> ```

nice test.


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68176)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 555 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiUv1l7mlo8YvuZ8li:-MfsaRFU70V9uY33qK8k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L555)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I would add maybe a a hint (errors.WithHint), or just expand the error message with "valid values are....."
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 629 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiW3Ng7o2axOiaOVho:-MfsaRiz7C4MX9syeOCE:b9j2neg) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L629)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

We already have that.  See running status (see changefeed_processors).
In order to update run status and indicate that we're pausing because of this policy, as opposed to user initiated action, we will probably need to export `pauseRequested` method  (as well as `onPauseRequestFunc`) so that we can directly call b.job.PauseRequested, and update run status in the passed in function.
</blockquote></details>

discussed offline. run status is updated internally in the progress column but is not visible in `SHOW JOBS` when paused.
___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 633 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiVcVaBOVWaSiTse4G:-Mfsaa1wEq0CZlmqs6dI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_stmt.go#L633)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> ```
> 		switch onError := changefeedbase.OnErrorType(details.Opts[changefeedbase.OptOnError]); onError {
> 		// default behavior
> 		case changefeedbase.OptOnErrorFail:
> 			return err
> 		// pause instead of failing
> 		case changefeedbase.OptOnErrorPause:
> 			// note: we only want the job to pause here if a failure happens, not a
> 			// user-initiated cancellation. if the job has been canceled, the ctx
> 			// will handle it and the pause will return an error.
> 			pauseErr := execCfg.JobRegistry.PauseRequested(ctx, jobExec.ExtendedEvalContext().Txn, jobID)
> 			if pauseErr != nil {
> 				return pauseErr
> 			}
> 			// TODO (ryan min): Populate pause reason with error once column is added (#67928)
> 		default:
> 			return errors.Errorf("unrecognized option value: %s=%s",
> 				changefeedbase.OptOnError, details.Opts[changefeedbase.OptOnError])
> 		}
> ```

I would move this into a helper `handleChangefeedError()` which returns an error (or nil)..
So, this code becomes:
```
if err != nil {
   err = handleChangefeedError(....)
}
return err
```
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 4265 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfiYjiE5qX2Ci44amLY:-MfsaaUC1cHOQwmJ255F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/66bc601cbd6dc9128cb00a47fd271a2fbe484264/pkg/ccl/changefeedccl/changefeed_test.go#L4265)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

require.NoError(t, feedJob.WaitForStatus...)?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 1 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-:-Mfsat-6EfiQQmGL0YWD:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 646 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfsaTjL6zxBQmcw65jc:-MfsaTjL6zxBQmcw65jd:bljnfo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/43a342ecac1a02d8061b83d6c23b264f70e0dc89/pkg/ccl/changefeedccl/changefeed_stmt.go#L646)):*
> ```Go
> 			// directly update running status to avoid the running/reverted job status check
> 			progress.RunningStatus = fmt.Sprintf("paused on error: %v", changefeedErr)
> 			log.Warningf(ctx, progress.RunningStatus)
> ```

nit: Let's be  more verbose here

```
log.Warningf(ctx, "job failed (%v) but is being paused because of on_error=pause", err)
```


<!-- Sent from Reviewable.io -->

spiffyy99 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68176)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @miretskiy and @spiffyyeng)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 646 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68176#-MfsaTjL6zxBQmcw65jc:-MfsflGX1NPvYOqXU7IL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/43a342ecac1a02d8061b83d6c23b264f70e0dc89/pkg/ccl/changefeedccl/changefeed_stmt.go#L646)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

nit: Let's be  more verbose here

```
log.Warningf(ctx, "job failed (%v) but is being paused because of on_error=pause", err)
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spiffyyeng commented:
TFTR! bors r+

spiffyy99 committed:
changefeedccl: add on_error option to pause changefeeds on failure

Previously, changefeeds always failed when encountering a non-
retryable error. This option allows the user to pause on failure
and resume later, while still failing as default behavior.

Resolves #67965

Release note (enterprise change): new 'on_error' option to pause
on non-retryable errors instead of failing.

spiffyyeng commented:
bors r+
